### PR TITLE
SE - Non-nullable value types: Keep NotNull constraint in branching

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #6105
~~❗ target branch is Martin/SE_6105_01_CheckNotNull of #6864~~

~~Needs to be retargeted once #6864 is merged.~~

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 1]
```diff
@@ -138,9 +138,9 @@ public void Branching_BoolExpression_LearnsBoolConstraint()
     Tag(""Value"", value);
 }";
             var validator = SETestContext.CreateCS(code).Validator;
-            validator.ValidateTag("True", x => x.HasConstraint<BoolConstraint>().Should().BeFalse());
-            validator.ValidateTag("False", x => x.HasConstraint(BoolConstraint.False).Should().BeTrue());
-            validator.ValidateTag("Value", x => x.HasConstraint(BoolConstraint.True).Should().BeFalse());    // Visited only for "true" condition
+            validator.ValidateTag("True", x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, BoolConstraint.True));
+            validator.ValidateTag("False", x => x.Should().HaveOnlyConstraint(BoolConstraint.False));
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:142]: Here we have another installment of the bug you mentioned.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 2]
```diff
@@ -64,24 +69,22 @@ private static ProgramState LearnBranchingConstraint<T>(ProgramState state, IBin
...
     private static SymbolicConstraint BinaryConstraint(BinaryOperatorKind kind, SymbolicValue left, SymbolicValue right)
     {
-        if (left is null && right is null)
-        {
-            return null;
-        }
-        else if (left is null || right is null)
+        var leftBool = left?.Constraint<BoolConstraint>();
+        var rightBool = right?.Constraint<BoolConstraint>();
+        if (leftBool is not null ^ rightBool is not null)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:74]: This is the same, isn't it? I find the `is null` easier to follow (positive logic)
```suggestion
        if (leftBool is null ^ rightBool is null)
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 3]
```diff
@@ -64,24 +69,22 @@ private static ProgramState LearnBranchingConstraint<T>(ProgramState state, IBin
 
     private static SymbolicConstraint BinaryConstraint(BinaryOperatorKind kind, SymbolicValue left, SymbolicValue right)
     {
-        if (left is null && right is null)
-        {
-            return null;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:69]: _Verification:_ Removing this short circuit has no functional change. It only deduplicates this `return null` with the `else return null`. And the new `?` down below are compensating for this.

Is that correct?

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Looks good, I'm requesting changes only to make sure my understanding is correct.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 4]
```diff
@@ -30,10 +30,15 @@ internal sealed class Binary : BranchingProcessor<IBinaryOperationWrapper>
     protected override SymbolicConstraint BoolConstraintFromOperation(ProgramState state, IBinaryOperationWrapper operation) =>
         BinaryConstraint(operation.OperatorKind, state[operation.LeftOperand], state[operation.RightOperand]);
 
-    protected override ProgramState LearnBranchingConstraint(ProgramState state, IBinaryOperationWrapper operation, bool falseBranch) =>
-        operation.OperatorKind.IsAnyEquality()
-            ? LearnBranchingConstraint<ObjectConstraint>(state, operation, falseBranch) ?? LearnBranchingConstraint<BoolConstraint>(state, operation, falseBranch) ?? state
-            : state;
+    protected override ProgramState LearnBranchingConstraint(ProgramState state, IBinaryOperationWrapper operation, bool falseBranch)
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:33]: >  It's just formatting, isn't it?

No. Previously we short-circuited when we learned about ObjectConstraint and didn't look into the BoolConstraint part. Now we visit both always.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs (hunk 2 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:74]: Yep. I think so. Tim was the first one who came up with xor for this kind of test on another PR, so we may ask him ;-).

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs (hunk 3 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:69]: Yes. That's right.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**